workspace: recover from missing operation

jonathantanmy opened the pull request:
In the commit message, I've added a discussion of alternate ways of how we can handle this situation, with my rationale for how I ended up doing it in the commit. If there are other (and maybe better) ways of handling this, feel free to let us know.

<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the branch rather than adding commits on top. Use force-push when
pushing the updated branch (`jj git push` does that automatically when you
rewrite a branch). Merge the PR at will once it's been approved. See
https://github.com/martinvonz/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

yuja commented on cli/src/cli_util.rs:
```diff
@@ -692,20 +692,81 @@ impl CommandHelper {
+                         Error message from read attempt: {e}"
+                    )?;
+                    (
+                        self.create_and_check_out_recovery_commit(ui, &mut workspace)?,
```
Perhaps, this fallback (and `create_and_check_out_recovery_commit()`) can be moved to the caller if this recovery method is an explicit option (something like `jj workspace update-stale --new-wc-commit`.)

It might be a bit inconvenient, but I think that's okay so long as the situation is rare.

jonathantanmy replied:
I was thinking about that too - I think that this whole function (for_stale_working_copy) should be moved to the caller, actually. cli_util.rs seems to be getting too large. If people have consensus, I can take a look at that.

yuja replied:
Ah yes, the original `for_stale_working_copy()` can also be moved to the caller.

Anyway, I prefer an explicit option to do `create_and_check_out_recovery_commit()` because the end result would look different, and it's valid to use that option to recover from the other stale situations.

jonathantanmy replied:
OK, I've refactored `for_stale_working_copy()` so it's now in the caller. As for the explicit option, since it's now moved to the caller, I don't think it's needed (because the caller always wants it).

yuja commented on cli/src/cli_util.rs:
```diff
@@ -692,20 +692,81 @@ impl CommandHelper {
+        mut_repo.set_wc_commit(workspace.workspace_id().clone(), new_commit.id().clone())?;
+        let ret = Ok(tx.commit("recovery commit"));
+
+        let mut locked_workspace = workspace.start_working_copy_mutation()?;
```
Might need to copy `check_working_copy_writable()` here, but I haven't tried.

jonathantanmy replied:
Just checking...do you mean that we should call `start_working_copy_mutation` before we create the new commit, so that if the working copy is not writable, we can print a message to the user without having written a useless commit?

`start_working_copy_mutation` already calls `check_working_copy_writable` (through `unchecked_start_working_copy_mutation`) so I don't think we need to call it here.

yuja replied:
> Just checking...do you mean that we should call `start_working_copy_mutation` before we create the new commit, so that if the working copy is not writable, we can print a message to the user without having written a useless commit?

No. (Doing that might be better, but isn't a requirement.)

> `start_working_copy_mutation` already calls `check_working_copy_writable` (through `unchecked_start_working_copy_mutation`)

This `workspace.start_working_copy_mutation()` doesn't go through `WorkspaceCommandHelper`. Perhaps, we can construct `WorkspaceCommandHelper` first by `for_loaded_repo()`/`workspace_helper_no_snapshot()`, and use `unchecked_start_working_copy_mutation()`. (The name "unchecked" is a bit odd, but it does some checks.)

jonathantanmy replied:
Ah, sorry for the wrong information (about `start_working_copy_mutation` calling `check_working_copy_writable` . I've implemented your suggestion. Also now we do call `for_loaded_repo()` before creating the commit, so there will be no useless commit if the working copy is not writable.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1917,14 +1986,18 @@ pub enum WorkingCopyFreshness {
-            WorkingCopyFreshness::Fresh | WorkingCopyFreshness::Updated(_) => false,
+            WorkingCopyFreshness::Fresh
+            | WorkingCopyFreshness::Updated(_)
+            | WorkingCopyFreshness::MissingOperation => false,
```
It's unclear whether `MissingOperation` should be considered "stale", but returning `true` will remove `!recovered` condition from the caller.

jonathantanmy replied:
Actually, if we return `MissingOperation` to where the caller (`cmd_workspace_update_stale`) checks `!recovered`, that can only happen during some sort of race, since `cmd_workspace_update_stale` already checks for that situation and then recovers. Maybe I should inline `is_stale` into `cmd_workspace_update_stale` first.

jonathantanmy replied:
I've inlined `is_stale` as I described and also treated `MissingOperation` differently.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1942,9 +2015,10 @@ pub fn check_stale_working_copy(
-            .read_operation(locked_wc.old_operation_id())?;
+        let wc_operation_data = match repo.op_store().read_operation(locked_wc.old_operation_id()) {
+            Ok(o) => o,
+            Err(_) => return Ok(WorkingCopyFreshness::MissingOperation),
```
nit: better to check if the error is `ObjectNotFound`

jonathantanmy replied:
OK.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -692,20 +692,81 @@ impl CommandHelper {
+        writeln!(
+            ui.stderr(),
+            "Created and checked out recovery commit {}",
+            new_commit.id().hex()
```
nit: use `short_commit_hash()`?

jonathantanmy replied:
OK.

jonathantanmy commented:
After some internal investigation at work, I've updated the code to also be more resilient when the current tree of the working copy is also missing (so, not only the operation). I've addressed all the comments (hopefully) except for the ones that involve some refactoring (moving `create_and_check_out_recovery_commit()` and some associated functions to the caller, and inlining `is_stale()`) - I'll continue to work on those.

yuja commented on lib/src/working_copy.rs:
```diff
@@ -108,6 +108,9 @@ pub trait LockedWorkingCopy {
     fn reset(&mut self, new_tree: &MergedTree) -> Result<(), ResetError>;
 
+    /// Update to the empty tree without touching the files in the working copy.
+    fn reset_to_empty(&mut self) -> Result<(), ResetError>;
```
Maybe `reset(new_tree)` could fall back to this mode if `old_tree` was missing.

@martinvonz any thoughts?

martinvonz replied:
That would probably also be fine. I kind of like being more explicit like this separate method since we don't expect the tree to ever be missing unless the operation is also missing. A missing tree would indicate a corrupt repo, which might be good to tell the user about. But I don't feel strongly either way.

yuja replied:
> I kind of like being more explicit like this separate method since we don't expect the tree to ever be missing unless the operation is also missing. A missing tree would indicate a corrupt repo, which might be good to tell the user about.

Yeah, that's a good point. OTOH, it's a bit weird that `reset_to_empty()` doesn't fail but `reset(empty_tree)` does. It might be a problem of function naming.

jonathantanmy replied:
I don't think `reset()` should do any falling back - it's called during normal operation whenever we check out, so if something unexpected happens, the user should know. Regarding naming, I purposely chose something that sounds like `reset()` so that implementers would know that both these functions need to do similar things, but if you can think of a better name, we can discuss it.

yuja commented on cli/src/commands/workspace.rs:
```diff
@@ -280,20 +281,73 @@ fn cmd_workspace_root(
+    let mut tx = repo.start_transaction(command.settings());
+    let mut_repo = tx.mut_repo();
+    let new_commit = mut_repo
+        .new_commit(command.settings(), vec![commit_id.clone()], tree_id.clone())
```
Does it mean the working copy commit will be diffed against the empty tree?

I think it's preferred to be based off the known tree if the tree still exists.

martinvonz replied:
Or we can set it to `commit.tree_id().clone()`. We know that exists at least (unless the repo is corrupt, of course).

jonathantanmy replied:
This is a recovery flow, so I wanted something that's as problem-free and clear to the user as possible (so not trying the known tree).

If we set it to `commit.tree_id().clone()` (as Martin suggests), would we also need to `reset()` to that tree? Even if we can, I think it's better to be consistent and use the empty tree everywhere.

jonathantanmy replied:
I don't think the selected tree matters for rebasing; it just changes the evolution of the commit (not sure whether jj uses a different term than "evolution", but what I mean is what you see when you run `obslog`).

Let's say we want to rebase this recovery commit. As far as I know, the only trees that matter are the commit's tree and the commit's parent's tree. As far as I can tell, what we're discussing is what the commit's predecessor's tree, which is not any of the trees that matter.

yuja replied:
> Let's say we want to rebase this recovery commit. As far as I know, the only trees that matter are the commit's tree and the commit's parent's tree.

Ah, that's correct. So my point applies if the recovery commit is "checked out", not "edited".

jonathantanmy replied:
The recovery commit is indeed "checked out" during the recovery process, so that does that mean that you still want to change the tree of the recovery commit (even though it will be immediately edited to another tree)?

martinvonz replied:
> The recovery commit is indeed "checked out" during the recovery process

Is that what we want? I would expect it to be edited. Otherwise we'd just create an empty recovery commit as parent for the actual recover commit. Or what am I missing?

martinvonz replied:
Wait, the next line below here calls `set_wc_commit()`, which is what we mean by "edit", so I think we're good. I'm instead confused about what the question was, but I haven't followed this discussion closely, so feel free to ignore me.

jonathantanmy replied:
> So my point applies if the recovery commit is "checked out", not "edited".
Ah, if you mean "...but since the recovery commit is both checked out and edited, my point no longer applies" then we are in agreement.

martinvonz replied:
Maybe the confusion is around terminology. What I assumed Yuya meant by "checked out " is in the `jj checkout/new` sense, i.e. that we create a new working-copy commit *on top* of some commit. Conversely, "edit" means to make the working copy point to the commit itself. Note that `jj new` is a combination of "create new commit on top of specified commits" and "edit the new commit" (without `--no-edit`). Does that clarify?

martinvonz replied:
> Does it mean the working copy commit will be diffed against the empty tree?
> 
> I think it's preferred to be based off the known tree if the tree still exists.

I agree. I would expect the initial recovery commit to either be empty or contain the changes in the working copy right away. For the latter, we would need an explicit call to snapshot here after resetting the working copy. That might be a little bit more work. I'm fine with either.

So, I agree that it doesn't matter much since it's barely observable to the user, but it still seems a little less surprising if the user does inspect it if we don't include an empty tree an a predecessor.

yuja replied:
> What I assumed Yuya meant by "checked out " is in the `jj checkout/new` sense, i.e. that we create a new working-copy commit _on top_ of some commit.

Yes. I misinterpret that there were the recovered base commit + snapshot from then. If the snapshot is non-empty, the working directory had uncommitted changes which the user might want to rebase or cherry pick.

jonathantanmy replied:
OK I think we are all fine with the current implementation then? If not, let me know exactly what you think I should change and I will try it.

yuja commented on cli/src/commands/workspace.rs:
```diff
@@ -280,20 +281,73 @@ fn cmd_workspace_root(
+                ))?,
+                false,
+            ),
+            Err(e) => {
```
We'll also need to test if `e` is of `ObjectNotFound` type.

But the user might want to continue the recovery in that situation. That's another reason why I think an explicit command option is better.

jonathantanmy replied:
I've added the test.

What do you mean by "continue the recovery"? The user has already initiated the recovery (by running "workspace update-stale") so presumably we don't need to ask the user whether they want to continue, since their intention is probably to do the recovery to the end.

yuja replied:
The user might want to try `update-stale` on some other kind of errors. If the error is `ObjectNotFound`, it's probably safe to continue. If the error isn't, `update-stale` might not be always the right way for recovery, but he can try.

jonathantanmy replied:
Ah, I see. So I think that adding the test (which I have done) gets what we want (if it's `ObjectNotFound`, it's safe and we can continue, but if not, we tell the user).

yuja replied:
> it's `ObjectNotFound`, it's safe and we can continue, but if not, we tell the user).

And the user might still want to try `update-stale` in that situation. That's what I wanted to say.

yuja commented on cli/src/commands/workspace.rs:
```diff
@@ -279,6 +280,22 @@ fn cmd_workspace_root(
+fn for_stale_working_copy(
+    command: &CommandHelper,
+    ui: &mut Ui,
+) -> Result<WorkspaceCommandHelper, CommandError> {
```
nit: reorder the arguments to `(ui, command)` like the other functions

jonathantanmy replied:
Done

yuja commented on cli/src/commands/workspace.rs:
```diff
@@ -306,22 +360,36 @@ fn cmd_workspace_update_stale(
+        // We should have already recovered from this situation above, so there must
+        // have been a concurrent operation.
+        {
+            return Err(user_error("Concurrent working copy operation. Try again."))
```
Another idea is to not map `OpStoreError` to `MissingOperation` by `check_stale_working_copy()`, and raise the original error here.

Then, `snapshot_working_copy()` attaches the hint if `check_stale_working_copy()` returned an Err.

yuja replied:
Because the error here means another weird thing happened?

I don't have strong opinion, but the underlying error might provide more context why the new operation got lost.

jonathantanmy replied:
Ah, I see. I was thinking that this is a similar case to WorkingCopyStale/SiblingOperation below, which doesn't report the underlying error but just says "Try again" in a weird state. I'll let @martinvonz make the call here.

jonathantanmy replied:
I checked with @martinvonz and I'll go with what @yuja says.

jonathantanmy replied:
I tried doing this, but this would require making WorkingCopyFreshness carry an error, which I think it can't (it derives PartialEq and errors are not PartialEq). If @yuja doesn't feel strongly about this, then hopefully the existing code is fine.

martinvonz replied:
> would require making WorkingCopyFreshness carry an error, which I think it can't (it derives PartialEq and errors are not PartialEq)

PartialEq is there only to make tests convenient. We have dropped it from errors before. The workaround is to use `matches!(err, MyError::MyVariant(_))` or similar.

(I haven't caught up with the rest of the thread yet)

martinvonz replied:
Having read through this, it sounds like Yuya's suggestion might simplify a little, so I think it's worth attempting it. If it gets complicated for some reason, feel free to leave it as is IMO (Yuya also seems fine with that).

yuja replied:
Since `check_stale_working_copy()` returns `Result<_, OpStoreError>`, maybe we can check if the error matches `ObjectNotFound` or not. It can also mean the ancestor operation was missing, but we'll probably want to do `update-stale` in that case.

jonathantanmy replied:
Thanks @yuja for the detailed instructions - I think I've done what you've suggested (making `check_stale_working_copy()` return the missing-operation case in the Err OpStoreError instead of Ok MissingOperation), and as @martinvonz says, it indeed makes things simpler.

yuja commented on cli/src/commands/workspace.rs:
```diff
@@ -306,22 +360,36 @@ fn cmd_workspace_update_stale(
-            "Nothing to do (the working copy is not stale)."
-        )?,
+        WorkingCopyFreshness::Fresh | WorkingCopyFreshness::Updated(_) => {
+            if !recovered {
```
Ah, okay, so the recovery should have already been done if the error was `MissingOperation`. Maybe better to return early in that case?

jonathantanmy replied:
Right now nothing happens after this `match`, so returning early has no effect.

I still don't want to return early, though. If we add code after this `match` (later), it would most likely need to take effect both in this recovery case and in other recovery cases (e.g. the working copy being stale).

yuja replied:
Then, the code can be structured as follows?
```rust
if !recovered {
    if check_stale_working_copy(...) {
    } else {
         locked_ws
            .locked_wc()
            .check_out(&desired_wc_commit)
            ...
    }
}
```

If I understand it, `recovered = true` means the `.check_out()` has been done.

jonathantanmy replied:
You are probably right that there is some redundancy, but I was thinking that it's better if each recovery step is as independent as possible. This is a recovery process so I think that predictability is much more important than efficiency.

yuja replied:
Well, the original recovery method is basically as follows:

1. Snapshot the current working copy changes on top of the "stale" operation
2. Merge diverged operations (which will rebase the new snapshot accordingly)
3. Now we're safe to modify the working copy, so update the working copy to the operation head's

(Beware that the step (3) depends on (1).)

The one implemented by this PR is:

1. Create new empty working copy commit for recovery
2. Reset the working copy to it (= the operation head's)
3. (Might be better to snapshot the working copy changes within this operation, but that will be done by the next `jj` invocation)

If I understand it, the recovery should have completed at the step (2), so it doesn't make sense to continue to the step (3) of the original method.

jonathantanmy replied:
> the recovery should have completed at the step (2), so it doesn't make sense to continue to the step (3) of the original method
Conceptually, I think that just because a recovery step was done doesn't mean that there is no longer anything to recover from (there might be more problems). It is true that in this case, after we recover, it is impossible for the working copy to be stale, but I don't think it's true in the general case.

Also, if we later add more recovery steps to `workspace update-stale`, I think it's clearer if each recovery step is allowed to run, rather than try to figure out which ones become redundant due to what happened earlier.

I think we need another tiebreaker vote on this (probably from @martinvonz ).

jonathantanmy replied:
I checked with @martinvonz and I'll go with what @yuja says.

jonathantanmy replied:
Instead of putting a lot of the function in a big `if !recovered` block, I checked `recovered` and exit early if that's the case, which I think is equivalent functionality-wise.

martinvonz commented on lib/src/local_working_copy.rs:
```diff
@@ -1752,6 +1762,18 @@ impl LockedWorkingCopy for LockedLocalWorkingCopy {
     }
 
+    fn reset_to_empty(&mut self) -> Result<(), ResetError> {
+        self.wc
+            .tree_state_mut()
+            .map_err(|err| ResetError::Other {
+                message: "Failed to read the working copy state".to_string(),
+                err: err.into(),
+            })?
```
We now have 5 copies (or very close) of this code. It's probably time to extract a method for it. Feel free to leave for a separate PR.

jonathantanmy replied:
OK, I'll leave for a separate PR.

martinvonz commented:
LGTM, but I'll let Yuya give the final approval. Thanks, both of you!

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1439,7 +1423,17 @@ Set which revision the branch points to with `jj branch set {branch_name} -r <RE
+                    ),
+                    e => e.into(),
+                },
+            )? {
```
nit: at your option, this can be organized as follows:
```rust
match check_stale_working_copy(locked_ws.locked_wc(), &wc_commit, &repo) {
    Ok(WorkingCopyFreshness::Fresh) => ...
    ...
    Err(OpStoreError::ObjectNotFound { .. }) => ...
    Err(err) => return Err(err.into()),
}
```

jonathantanmy replied:
Thanks, I've used what you suggested.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -1980,9 +1963,10 @@ pub fn check_stale_working_copy(
+        let wc_operation_data = match repo.op_store().read_operation(locked_wc.old_operation_id()) {
+            Ok(o) => o,
+            Err(e) => return Err(e),
+        };
```
this change is no longer needed

jonathantanmy replied:
Ah, yes. I've reverted the change.

yuja commented on cli/src/commands/workspace.rs:
```diff
@@ -279,6 +282,76 @@ fn cmd_workspace_root(
+    let workspace_id = workspace.workspace_id().clone();
+    let tree_id = repo.store().empty_merged_tree_id();
+
+    let mut workspace_command = command.for_loaded_repo(ui, workspace, repo.clone())?;
```
This is basically the same as `command.workspace_helper_no_snapshot()`?
Can be something like
```rust
    let mut workspace_command = command.workspace_helper_no_snapshot(ui)?;
    let workspace_id = workspace_command.workspace_id().clone();
    let mut tx = workspace_command.start_transaction().into_inner();
```

jonathantanmy replied:
Ah, good catch. I used your suggestion, except that for the last line, I had to use `repo.start_transaction(command.settings())` instead to avoid borrow checker problems (before this line, I call `workspace_command.unchecked_start_working_copy_mutation()`, which also borrows `workspace_command` mutably).

yuja replied:
It's a bit tricky, but these lines can be reordered. `.start_transaction().into_inner()` releases `&mut workspace_command`.

`repo.start_transaction()` doesn't record the command arguments in the operation. You can see the difference in `jj op log`.

jonathantanmy replied:
Ah, I did not realize this. Let me send a PR to correct this.

yuja commented on cli/src/commands/workspace.rs:
```diff
@@ -279,6 +282,76 @@ fn cmd_workspace_root(
+    let ret = Ok(tx.commit("recovery commit"));
+
+    locked_workspace.locked_wc().reset_to_empty()?;
+    locked_workspace.finish(operation.id().clone())?;
```
Since `tx.commit()` creates new operation, its id should be recorded instead.
```rust
    let repo = tx.commit("recovery commit");
    ...
    locked_workspace.finish(repo.op_id().clone())?;
```

The operation id recorded here will be overwritten by `maybe_snapshot()` though.

jonathantanmy replied:
Ah, thanks for noticing. I've changed it.

yuja commented on cli/src/commands/workspace.rs:
```diff
@@ -279,6 +282,76 @@ fn cmd_workspace_root(
+    command: &CommandHelper,
+    ui: &mut Ui,
+    workspace: Workspace,
+) -> Result<Arc<ReadonlyRepo>, CommandError> {
```
nit: `(ui, command)` instead of `(command, ui)`?

(The `workspace` parameter can be removed by switching to `workspace_helper_no_snapshot()`.)

jonathantanmy replied:
Done.

yuja reviewed (approved):
Thanks!

jonathantanmy commented:
Thanks everyone. I'll go ahead and merge this PR.
